src/pq_flash_index.cpp

```diff
@@ -1536,6 +1536,85 @@ template <typename T, typename LabelT> diskann::Metric PQFlashIndex<T, LabelT>::
     return this->metric;
 }
 
+template <typename T, typename LabelT> std::vector<T> PQFlashIndex<T, LabelT>::get_vector(std::uint64_t VID)
+{
+    ScratchStoreManager<SSDThreadData<T>> manager(this->thread_data);
+    auto this_thread_data = manager.scratch_space();
+    IOContext &ctx = this_thread_data->ctx;
+    std::vector<AlignedRead> read_reqs;
+    AlignedRead read;
+    char *buf = nullptr;
+    alloc_aligned((void **)&buf, SECTOR_LEN, SECTOR_LEN);
+    read.len = SECTOR_LEN;
+    read.buf = buf;
+    read.offset = NODE_SECTOR_NO(VID) * SECTOR_LEN;
+    read_reqs.push_back(read);
+    reader->read(read_reqs, ctx);
+
+#if defined(_WINDOWS) && defined(USE_BING_INFRA)
+    if ((*ctx.m_pRequestsStatus)[0] != IOContext::READ_SUCCESS)
+    {
+        aligned_free(buf);
+        return std::vector<T>();
+    }
+#endif
+
+    char *node_buf = OFFSET_TO_NODE(buf, VID);
+    T *node_coords = OFFSET_TO_NODE_COORDS(node_buf);
+
+    std::vector<T> out(get_data_dim());
+    memcpy(out.data(), node_coords, sizeof(T) * get_data_dim());
+
+    return out;
+}
+
+template <typename T, typename LabelT>
+std::vector<std::uint8_t> PQFlashIndex<T, LabelT>::get_pq_vector(std::uint64_t VID)
+{
+    std::vector<std::uint8_t> out(this->n_chunks);
+    std::uint8_t *pq_vec = &this->data[VID * this->n_chunks];
+    memcpy(out.data(), pq_vec, this->n_chunks);
+    return out;
+}
+
+template <typename T, typename LabelT>
+std::vector<std::uint64_t> PQFlashIndex<T, LabelT>::get_neighbors(std::uint64_t VID)
+{
+    ScratchStoreManager<SSDThreadData<T>> manager(this->thread_data);
+    auto this_thread_data = manager.scratch_space();
+    IOContext &ctx = this_thread_data->ctx;
+    std::vector<AlignedRead> read_reqs;
+    AlignedRead read;
+    char *buf = nullptr;
+    alloc_aligned((void **)&buf, SECTOR_LEN, SECTOR_LEN);
+    read.len = SECTOR_LEN;
+    read.buf = buf;
+    read.offset = NODE_SECTOR_NO(VID) * SECTOR_LEN;
+    read_reqs.push_back(read);
+    reader->read(read_reqs, ctx);
+
+#if defined(_WINDOWS) && defined(USE_BING_INFRA)
+    if ((*ctx.m_pRequestsStatus)[0] != IOContext::READ_SUCCESS)
+    {
+        aligned_free(buf);
+        return std::vector<std::uint64_t>();
+    }
+#endif
+
+    char *node_buf = OFFSET_TO_NODE(buf, VID);
+    std::uint32_t *nbr_buf = OFFSET_TO_NODE_NHOOD(node_buf);
+    std::uint64_t nnbrs = (std::uint64_t)(*nbr_buf);
+
+    std::vector<std::uint64_t> out(nnbrs);
```

Comment:
yes, i think it might be more efficient than relying on return value optimization (as it depends on caller pattern and may not always be guaranteed to avoid copying)?